### ISSUE TITLE
CSV export - date and time fields are in excel format

### ISSUE THREAD

**Issue by oberhm** [abap2xlsx/abap2xlsx#1134]
Dear all,

we use the "CSV Writer" class to transfer a file.
In the source table we have date and time fields.

The CSV Writer exports the fields in excel format.
e.g. Date 17.08.2023 => 45155

We found a possible solution for that:
When the field lv_data_type is set in the method SET_CELL of the class ZCL_EXCEL_WORKSHEET to "D" or "T" the exporter class handles the fields correct.

If this might by a solution, I will create a pull request

thanks 
Matthias

**Comment by sandraros** [comment]
Sorry for the delay. Are you sure there is a bug in the CSV writer? Could you post a minimal reproducible example so that we can check ourselves? Thank you!

**Comment by oberhm** [comment]
[z_abap2xlsx_issue_1134.txt](https://github.com/abap2xlsx/abap2xlsx/files/12697525/z_abap2xlsx_issue_1134.txt)

Here is an example for the problem.
In XLSX download the date field is set with 45155, which is converted by Excel to real date 17.08.2023. In the csv export the value is also set to 45155, which is no date. If the csv is opened without excel the field is a number and no date.

I think in csv export, the date field should by a date in real format (not the excel internally format).

**Comment by sandraros** [comment]
Thank you. I agree 45155 is buggy in CSV. Your program will help me analyze, and I'll confirm after week-end, based on your latest comments, probably I'll agree on your PR and merge it.

**Comment by oberhm** [comment]
Hi,
after your comment yesterday I debugged again.
Maybe another solution could help, but here I'm not familiar with the data which abap2xlsx is using.

in the method CREATE_CSV of class ZCL_EXCEL_WRITER_CSV there is a coding which is called "Retrieve Cell Style format and data type"
The field cell_style is filled in the structure but this coding did not fill the variable ls_numfmt. 
Maybe we get here a correct style then the data_type is set here and must not be set in the method SET_CELL.

**Comment by darnoc312** [comment]
This is my suggestion:
The NUMFMT Field is retrieved from the dynamic styles. (excel->get_styles_iterator( ) and following WHILE Loop)

At the begining of zcl_excel_writer_2007->create( ) are the following lines

* Start of insertion # issue 139 - Dateretention of cellstyles
  me->excel->add_static_styles( ).
* End of insertion # issue 139 - Dateretention of cellstyles

to create a dynamic style for any static style if not done so far.
Else some dynamic styles could be missing.

This method call is still not used at the begining of zcl_excel_writer_csv->create( ) and should be included there.

**Comment by oberhm** [comment]
Hi All,

I changed the coding like darnoc mentioned.
This will solve my problem.

But nevertheless: Maybe for an future release. we should check, why the field data_type is not set correctly.

Happy to hear comments.

Matthias

**Comment by darnoc312** [comment]
Data_type is set correctly,

This field is the Excel data type with the following valid values:

![image](https://github.com/abap2xlsx/abap2xlsx/assets/135710507/e3c47ff4-306f-4bf6-b73e-de649736014e)
From: https://itecnote.com/tecnote/excel-what-do-excel-xml-cell-attribute-values-mean/

Currently in abap2xlsx only space and 's' are used (space leads to default 'n').

The settings for this field in zcl_excel_writer_csv->create_csv differ from this list.
These are local settings for local purpose in a local copy of table zcl_excel_worksheet->sheet content and perhaps they look like ABAP Inttype (And the query for upper case letters seems to be unnecessary).
But if the destination is a XML Output the value of data type set by  zcl_excel_worksheet->set_cell must be according to the list above.

**Comment by sandraros** [comment]
Starting the review. Posting another minimal example to reproduce + testing all supported ABAP types:
```abap
REPORT z_abap2xlsx_issue_1134.
DATA gc_save_file_name TYPE string VALUE 'abap2xlsx_issue_1134.csv'.
INCLUDE zdemo_excel_outputopt_incl.
PARAMETERS: p_csv  RADIOBUTTON GROUP rad1 DEFAULT 'X',
            p_xlsx RADIOBUTTON GROUP rad1.
START-OF-SELECTION.
  TRY.
      DATA(lo_excel) = NEW zcl_excel( ).
      DATA(lo_worksheet) = lo_excel->get_active_worksheet( ).
      types ty_ref_to_values type standard TABLE OF ref to data with EMPTY KEY.
      LOOP AT VALUE ty_ref_to_values( ( REF #( 'Hello' ) )
                                      ( NEW d( '20230930' ) )
                                      ( NEW decfloat16( '-3.14' ) )
                                      ( NEW f( '.3' ) )
                                      ( NEW int1( 3 ) )
                                      ( NEW int2( -314 ) )
                                      ( REF #( -314 ) )
                                      ( NEW sflight-price( '314.16' ) )
                                      ( REF #( `Hello` ) )
                                      ( NEW t( '151416' ) )
                                      ( NEW syhex02( 'FFFE' ) )
                                    ) INTO DATA(ref_to_value).
         ASSIGN ref_to_value->* TO FIELD-SYMBOL(<value>).
         lo_worksheet->set_cell( ip_column = zcl_excel_common=>convert_column2alpha( ip_column = sy-tabix + 0 ) " pass by value
                                 ip_row    = 1
                                 ip_value  = <value> ).
      ENDLOOP.
      DATA(lv_writerclass_name) = COND #( WHEN p_csv = 'X' THEN 'ZCL_EXCEL_WRITER_CSV' ).
      IF p_xlsx = 'X'.
        REPLACE '.csv' IN gc_save_file_name WITH '.xlsx'.
      ENDIF.
      lcl_output=>output( cl_excel            = lo_excel
                          iv_writerclass_name = lv_writerclass_name ).
    CATCH cx_root INTO DATA(lx_root).
      MESSAGE lx_root TYPE 'I' DISPLAY LIKE 'E'.
  ENDTRY.
```
Current bug:
![image](https://github.com/abap2xlsx/abap2xlsx/assets/34005250/67edd7c3-a1f3-4d6f-81d5-73d9e5041f3e)

After applying the proposed fix (#1140):
![image](https://github.com/abap2xlsx/abap2xlsx/assets/34005250/c8c01d40-af38-4f27-a44f-9eefa2504b03)

For information, XLSX format:
![image](https://github.com/abap2xlsx/abap2xlsx/assets/34005250/fe66542b-4044-44cc-8a29-d962067f411c)
